File: algo_10_greedy/L605.cpp
//
// Created by 李恒 on 2020/7/3.
//
//种花问题

#include <iostream>
#include <vector>

using namespace std;
// 一次遍历，同时看前后是否为0
class Solution {
public:
    bool canPlaceFlowers(vector<int>& flowerbed, int n) {
        int len = flowerbed.size();
        int cnt = 0;
        for (int i = 0; i < len && cnt < n; i++) {
            if (flowerbed[i] == 1) {
                continue;
            }
            int pre = (i==0 ? 0:flowerbed[i-1]);
            int next = (i==len-1 ? 0:flowerbed[i+1]);
            if (pre==0 && next==0) {
                cnt++;
                flowerbed[i] = 1;
            }
        }
        return cnt >= n;
    }
};

File: python-algorithm/algo_10_greedy/L605.py
# -*- coding: utf-8 -*-


# 一次遍历，同时看前后是否为0
class Solution:
    def canPlaceFlowers(self, flowerbed: List[int], n: int) -> bool:
        sz = len(flowerbed)
        cnt = 0
        for i in range(sz):
            if flowerbed[i] == 1:
                continue
            pre = (0 if i==0 else flowerbed[i-1])
            nxt = (0 if i==sz-1 else flowerbed[i+1])
            if pre==0 and nxt==0:
                cnt += 1
                flowerbed[i] = 1 # 容易漏掉
        return cnt >= n
